File: ts/grant.ts
import { parse } from "./parser";
import { simSearch, retrieveDataFromLlm, determineSchema } from "./rag";
import { OllamaEmbeddings } from "@langchain/community/embeddings/ollama";
import { MemoryVectorStore } from "langchain/vectorstores/memory";
import * as fs from "fs";
import * as path from "path";

async function collectPdfFiles(dir: string): Promise<string[]> {
  let files: string[] = [];
  const entries = await fs.promises.readdir(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files = files.concat(await collectPdfFiles(fullPath));
    } else if (entry.isFile() && entry.name.toLowerCase().endsWith(".pdf")) {
      files.push(fullPath);
    }
  }
  return files;
}

const args = process.argv.slice(2);
if (args.length < 1) {
  console.log(
    "Usage: ts-node grant.ts <k-value> [--folder <dir> | <pdf1> [pdf2 ...]]\n" +
      "Example: ts-node grant.ts 3 --folder ./pdfs"
  );
  process.exit(1);
}

const k = parseInt(args[0], 10);
let folder: string | undefined;
const filepaths: string[] = [];
for (let i = 1; i < args.length; i++) {
  const arg = args[i];
  if ((arg === "--folder" || arg === "-f") && i + 1 < args.length) {
    folder = args[i + 1];
    i++;
  } else {
    filepaths.push(arg);
  }
}

async function resolveFilepaths(): Promise<string[]> {
  if (folder) {
    return collectPdfFiles(folder);
  }
  if (filepaths.length > 0) {
    return filepaths;
  }
  console.error(
    "Error: must specify either --folder <dir> or one or more PDF files"
  );
  process.exit(1);
}

(async () => {
  console.log("Loading PDFs...");
  const filesToParse = await resolveFilepaths();
  const pages = await parse(filesToParse);

  const embeddings = new OllamaEmbeddings({ model: "nomic-embed-text" });
  const vectorStore = await MemoryVectorStore.fromDocuments(pages, embeddings);

  const llmQueries: Record<string, string> = {
    general: "What is the full name of the grant? List all projects or programs stated, including each project's name, start date, and end date.",
    spending: "What is the total grant amount? Break down the spending into: fringe benefits, indirect costs, travel, equipment, and other. For \u201cother,\u201d provide a list of items with their names and cost. Return only valid JSON."
  };

  const vecQueries: Record<string, string> = {
    general: "Information about the grant's official name, and the names, start and end dates of any funded projects or programs.",
    spending: "Details about total funding, and how the grant money is allocated \u2014 including fringe benefits, indirect costs, travel, equipment, and other types of spending."
  };

  let grant: any = {};

  for (const key of Object.keys(llmQueries)) {
    console.log("Similar searching...");
    const vecQuestion = vecQueries[key];
    console.log("Q:", vecQuestion);
    const context = await simSearch(vecQuestion, k, vectorStore);
    const schema = determineSchema(key);
    console.log("Got context");

    console.log("Asking LLM...");
    const llmQuestion = llmQueries[key];
    console.log("Q:", llmQuestion);
    const response = await retrieveDataFromLlm(llmQuestion, context, schema);
    console.log("response:", response);

    console.log("Updating grant json...");
    grant = { ...grant, ...response };
    console.log(JSON.stringify(grant));
  }

  await fs.promises.writeFile("grant.json", JSON.stringify(grant, null, 4));
})();

File: grant.py
'''
    main script that generates json of grant data
'''

from parser import parse
from rag import sim_search, determine_pyobj, retrieve_data_from_llm
from langchain_ollama import OllamaEmbeddings
from langchain_core.vectorstores import InMemoryVectorStore
import argparse
import json
import os
import re


def main() -> None:
    argparser = argparse.ArgumentParser(description="Parse grant PDF files")
    argparser.add_argument(
        "files",
        nargs="*",
        help="Path(s) to one or more PDF files for the same grant",
    )
    argparser.add_argument(
        "-k",
        type=int,
        default=4,
        help="# nearest neighbors to retrieve for similarity search",
    )
    argparser.add_argument(
        "-f",
        "--folder",
        help="Folder to recursively search for PDF files",
    )
    args = argparser.parse_args()

    file_list = []
    if args.folder:
        if not os.path.isdir(args.folder):
            argparser.error(f"Folder '{args.folder}' does not exist")
        for root, _, files in os.walk(args.folder):
            for fname in files:
                if fname.lower().endswith(".pdf"):
                    file_list.append(os.path.join(root, fname))
    if not file_list and args.files:
        file_list = args.files
    if not file_list:
        argparser.error("Please specify a folder or one or more PDF files")

    print("Loading PDF(s)...")
    print(file_list)
    try:
        pages = parse(file_list)
    except Exception as e:
        joined = ", ".join(file_list)
        print(f"Failed to load PDF(s) '{joined}': {e}")
        return

    print("Starting embedding")
    embeddings = OllamaEmbeddings(model="nomic-embed-text")
    vector_store = InMemoryVectorStore.from_documents(pages, embeddings)

    llm_queries = {
        "general": (
            "What is the full name of the grant? List all projects or programs stated, including each project's name, start date, and end date."
        ),
        "spending": (
            "What is the total grant amount? Break down the spending into: salary, fringe/payroll benefits, indirect costs, travel, equipment, and other. For “other,” provide a list of items with their names and cost. Return only valid JSON."
        ),
    }
    vec_queries = {
        "general": (
            "Information about the grant's official name, and the names, start and end dates of any funded projects or programs."
        ),
        "spending": (
            "Details about total funding, and how the grant money is allocated — including salary, fringe/payroll benefits, indirect costs, travel, equipment, and other types of spending."
        ),
    }

    grant_json: dict[str, object] = {}

    # go through each batch of queries to efficiently/accurately get the proper data from pdf
    for query in llm_queries:
        print("Similar searching...")
        vec_question = vec_queries[query]
        print("Q: ", vec_question)
        context = sim_search(vec_question, args.k, vector_store)
        py_obj = determine_pyobj(query)
        print("Got context")

        print("Asking LLM...")
        llm_question = llm_queries[query]
        print("Q: ", llm_question)
        response = retrieve_data_from_llm(llm_question, context, py_obj)
        print("response: ", response)

        print("Updating grant json...")
        x = json.loads(json.dumps(grant_json))
        x.update(response)
        print(json.dumps(x))
        grant_json = x

    # write to json file
    obj = json.dumps(grant_json, indent=4)
    model_name = os.getenv("MODEL", "model")
    if args.folder:
        base = os.path.basename(os.path.normpath(args.folder))
    else:
        # join pdf file names without extensions
        names = [os.path.splitext(os.path.basename(f))[0] for f in file_list]
        base = "_".join(names)
    base = re.sub(r"[^A-Za-z0-9_-]+", "_", base)
    filename = f"grant-{model_name}-{base}.json"
    with open(filename, "w") as outfile:
        outfile.write(obj)
    print("Finished writing to "+filename)


if __name__ == "__main__":
    main()
